### ISSUE TITLE
Generated project customization support

### ISSUE THREAD

**Issue by masterspline** [build2/build2#16]
I think making build2 easier to use, user friendly and lowering entry threshold is very very very important to make it widely used. My suggestion is to make cxx and hxx extensions not single for each type but all widely used by users (and make this by default)
```
hxx{*}: extension = h, hpp, hxx
cxx{*}: extension = cc, cp, cxx, cpp, c++, C
```
Reason is I personally hate cxx as c++ file extension, Boris has other preferences. Header files for c++ is better to use .h extension (for me). They are easier to distinguish when in same directory exist files with same names
```
file.hpp
file.cpp
// vs
file.h
file.cpp
```
(and in present every C only file MUST have 

```
#ifdef __cplusplus
namespace project_namespace
{
  extern "C"
  {
#endif
```
guards to be includable both in C and C++). But this is my opinion. And of course different people have different preferences (for file extensions too).

Simple solution is to have multiple extensions for file types. In such case 
```
exe{util}: {hxx ixx txx cxx}{** -version}
```
will compile files with all matched extensions (so everybody can use any preferred extensions or even mix them in single project without any configuration, because by default all extensions will match). In case `hxx{version}: in{version} $src_root/manifest` first (default) extension can be used for generated file.
Configuring
```
hxx{*}: extension = hxx
cxx{*}: extension = cxx
```
will return current behavior.

**Comment by boris-kolpackov** [comment]
Thanks for writing down your thoughts. While I can see the "solution" you are proposing, I am not clear from your description about what the "problem" is in specific terms (e.g., "I've done this and that, the result is wrong/counter-intuitive/etc.").

My view of the (admittedly unfortunate) situation is that yes, we don't have a standard C++ header/source file extension (and are probably unlikely to ever get one). It is, however, pretty easy to configure your preferred extension on the per-project basis in `build/root.build`. Plus, the `bdep-new(1)` command supports creating projects with a few commonly-used naming schemes (currently `.?xx` and
`.?pp` but we could add more if necessary). Finally, projects that use multiple file extensions are (and should be) uncommon. So given this view, the current `build2` semantics seems to be adequate.

While I don't want to go into a detailed discussion/analysis of your solution before having a clear and specific understanding of the problem, one thing to note at the outset is that pattern-matching multiple extensions will be punishing performance-wise, especially on "lesser" platforms.

Also, let me add a link to the discussion of the C++ header/source extensions that disagrees (with rationale) with a few of your points (like reusing `.h` extension for C++ headers):

https://build2.org/build2-toolchain/doc/build2-toolchain-intro.xhtml#proj-struct-src-name

**Comment by masterspline** [comment]
Detailed description
====
I am trying to solve problem "build2 is too complex to use". It MUST be user friendly or it will be dead. I will show in example.

Cmake user story: when I start new project I copy existing CMakeLists.txt file and modify it (single file!) or create new one (`cmake_minimum_required()`, `project()`, `add_executable()`), then `mkdir src include build`. Done. Easy.

`build2` have more functionality (package management), so it will be more complex. But it is too complex. When I start new project (even for simple test program to check how something work) I can't just copy and edit file, I need to use `bdep new ...` and I need to remember command line parameters (but it is impossible, so I need to remind them, so get `man bdep`, find example, copy-paste-edit-run it). Then I need to edit `build/root.build`, because I prefer `.cpp` scheme (I can use `bdep new` to generate `.cpp` scheme, but I need to not forget to do this, so it is easy to fix config later, because I do not need to remember syntax, probably this is easier for me only). Also I prefer `.h` extension for header files (probably this will change sometimes), so I MUST edit config. Then start fighting with IDE for it works with my project (with cmake I do not need this, and of cause this is need to be fixed on IDE side). In short, this long story is just to show that there are huge amount of difficulties when working with build2.

And do not care some troubles out of `build2` scope, build2 must be as easy to use (providing required functionality) as possible to be used. Currently conan and vcpkg fighting for place under sun. Both want to be standardized (by cpp committee!) and both have big company with big resources. build2 will be convenient or dead. My feature request make build2 suitable for any file naming preferences by default, so `bdep new` command line argument can be removed (config editing will be enough). So (almost) every user even not need to know about this config option. This make build2 easy to use and leave possibility for current behavior.

> pattern-matching multiple extensions will be punishing performance-wise, especially on "lesser" platforms.

As I see this situation, for pattern match You need to compare with single string or You need to look in radix tree (or even sequential search for 3 byte string converted to uint32_t by lookup in 256bit SSE register in uint32_t array, 8 extensions at once, do not think extension list will be longer). Not big difference. And of cause as project author You know better. As I understand, current choice is make build2 slightly slower in pattern matching or make configuration slightly inconvenient (manual). My choice - convenience. Let machine work.

> Also, let me add a link to the discussion of the C++ header/source extensions that disagrees

This is Your opinion. I read it. Disagree. I do not understand, why convenient build tool need to force user care of headers (actually project headers behave exactly the same as system headers). To build project user must show which `.cpp` files to compile for given target, if build system needs info about headers - ask compiler, they are included in `.cpp` file (gcc -M or some other) and filter system headers from project headers. Compiler need to care about headers used by translation unit, IDE need to care of headers (to show in project file list, but again they need to ask for compiler about headers). I do not know why build system need info about headers.
> There can be a need for both C and C++ headers for the same module.

build system MUST be ready for same `.h` header will be used in C and in CPP translation unit
> It allows tools to accurately determine the language from the file name.

Language configured for build target, so there is dependency target->.cpp/.c->.hpp/.h. Headers commonly used in cpp and c project in same time, for .c files it is not common but possible too.
> It is easier to search for C++ source code using wildcard patterns (*.?pp).

As I mentioned above difference is not huge and let machine work but human creates.

Also some offtopic. C++ is not Haskell or Rust. There is no right way for some task. Any task can be solved by different ways and in C++ different people prefer different solutions. So good tool need to support this different ways to solve task. But today is not 90-th, there are many tools for any task - so Your tool need to be easy to use/user friendly/convenient or it will unused.

Short version
====
build2 have many difficulties/troubles in usage. For people use tool Modern tool must be easy to use. For many people use C++ tool it need support different ways to solve task. My feature request eliminate need to configure files extensions for (almost) everybody, so any developer with his own preferences can use default provided configuration in any project even not knowing about it. The price - tiny performance decrease in pattern matching (using sequential search or radix tree in CPU cache line).

**Comment by boris-kolpackov** [comment]
Thanks for providing more information and especially the description of your workflow and the difficulties you are experiencing. I also appreciate your passion for making `build2` successful. However, I also think we must keep "cool heads" and not rush into any quick solutions, especially if this is a sign of a larger/more fundamental problem (I am referring to [your other issue](https://github.com/build2/bdep/pull/1)).

So, based on your description, my summary of the overall problem is that creating new `build2` projects is too difficult, especially if the desired settings do not match the default. And the "copy one of the existing projects" approach does not work well because the `build2` build/package information is spread over multiple files that would require changing. Do you think this is accurate?

Let me first discuss the "copy existing project" issue: I don't think the "single file" approach (found in CMake, etc) will work for `build2`. While having things spread over multiple files might be inconvenient for really simple projects, there are good reasons why things are organized this way and which become apparent when one deals with more complex projects. In other words (and as you have said yourself), `build2` provides more functionality and is geared towards handling more complex projects gracefully which, unfortunately, necessitates a more elaborate, multi-file project structure. 

So, in this light, I believe `bdep-new` is our only viable way forward. And I think Rust's experience shown that it can work. The question is then how can we fix it to be more user-friendly?

From the above description and your other issue I see the following problems:

1. Inability to specify `.h/.cpp` naming scheme (and thus having to edit the generated project).
2. Even if there were a way, having to remember the necessary options and specify them every time is tedious.
3. Inability to customize the project post-creation (e.g., for IDE support, etc).

Solving (1) is pretty easy: we can add support for customizing individual extensions, something along these lines:

```
bdep new -l c++,hxx=h,cxx=cpp
```

While (2) is harder, it feels like having persistent configuration (a-la `git config --global`) could be a reasonable solution. It's not perfect since it requires an extra step for a new user before things become smooth and easy, but it's better than nothing. And it is general-purpose: you can also configure default language, whether you want tests/unit-tests, etc (we should probably make it project type (`-t`) and language (`-l`) specific so you can have different defaults). And once you've done this, creating new projects exactly the way you want them becomes as easy as:

```
bdep new -t exe hello
bdep new -t lib libhello
```

For (3) I think the project template idea I've described [here](https://github.com/build2/bdep/pull/1) is the best approach. And it fits with the global configuration idea quite nicely: you can specify your template there once and have it applied every time your create a new project. Another idea I had is a post-creation hook where you can have `bdep-new` execute an arbitrary command that can further customize/initialize a project (say, call and IDE to create something, etc).

What do you think?

**Comment by NikolausDemmel** [comment]
> 2\. Even if there were a way, having to remember the necessary options and specify them every time is tedious.

Just a though while reading this. While a config file to customize defaults is probably a good idea, and alternative (or additional) way to support the casual user that doesn't remember all the options would maybe be some interactive- or wizard-mode for `dbep new`, that interactively asks a bunch of questions for various options like name, type, dependencies, file extensions, ... (showing the default and possibly other popular alternatives), or shows a menu-like list of the most important options, their current defaults and possibly popular alternative values and gives the user a chance to change any one of them. In the end it would execute the appropriate command (and also show the command line to the user, so that after a few runs with the wizard the user can maybe easily remember the command line syntax.

**Comment by boris-kolpackov** [comment]
Yes, that's a good idea (I've also thought of an interactive `bdep-new` mode but forgot to mention it). I especially like the idea of showing the non-interactive command line equivalent.

**Comment by masterspline** [comment]
> bdep new -l c++,hxx=h,cxx=cpp

It's easier to edit configs after, then remember this command line. So current (existing) way is easier.

Global AND user configs with templates are looks useful. Interactive mode are too. But I think text configs are more convenient because it You need to fix two options You will need to edit only two lines. Interactive mode will ask too many questions even in case You can choose questions groups (and also because I am ex sysadmin and I prefer edit configs :). Also I think this config better to make explicit (used by --config `path/to/file.conf`, not as for git). In this case user can have template for config somewhere (or distributive provided), copy it in workdir, edit, then run `bdep new`. This way will be slightly more easy to use then run `bdep new`, then (after) edit generated files (but not much easier). This will work for creating `.gitignore` too, but we have 3 OSes, 3 IDEs and many different tools, so (I think) QtCreator, IDEA and VS `.gitignore`s better to be included in distributive, other tools can be configured manually or from user configs.

For this feature request I think my solution (only for this task) is easier to use and more convenient because it totally eliminates needs for configuration at all and suitable for (almost) everybody. Just allow multiple extensions and only in exotic cases people will need to fix generated configs.

P.S. I can write request for `bdep.conf` for discussion of config solution.

**Comment by Klaim** [comment]
> Global AND user configs with templates are looks useful.

Suggestion: I think a directory config (searched in upper directories until) would be more useful than a global one.  Like how clang-format works.

**Comment by masterspline** [comment]
Some offtopic (philosophy)
====
There are two ways to make things in software development (tools, projects, making solutions): the Right way an easy way.

Right way have strong philosophy with solid super Idea behind it. In this way acceptable only solutions obey this Idea. All other things are wrong. This way leads to niche products (ex.: haskell and all functional programming) or too difficult to use niche products (ex.: rust). Because in general programming there are too many different tasks requires different tools and too right tool not fit wide rage of tasks but good in it's niche. And such solution is perfect (all obey strong philosophy) and work well in his niche but useful only in this niche. I called this way fanatics way.

Easy way have some rules which is more like orienting point, but the main joining idea is easiness to use for real life problems (haskell created by mathematics for mathematics).This is C++, Java, Kotlin, Golang, Python way. Language provides wide rage of easy to use tools for different real life tasks. They all fit each other to make single monolith language (but not such solid as in first way). This way leads to not perfect solutions (different tools are not perfectly fits each other) but useful in real life problems. C++ is complex because it have too many tools but each tools are needed for solving some task, remove it and C++ become useless for this tasks. C++ have strange syntax, sometimes scare and this is because all different C++ tools must have monolith syntax. I think this is only suitable way in software development.

Why did I wrote this. I thin build2 have too strong ideology. This do not work in c++. Embedded and distributed developers use different solutions for (seemingly) same tasks. And there are many people have different preferences. So build2 need to be flexible and support different people with different preferences. Also it must be convenient (there are other tools, but personally I think build2 can become best if will move faster then other in right direction). Right for me is easy to use tool solving my tasks and I do not care which ideology is behind this tool, that is why I think the only solution is easy to use solution. The "Right" solution do not work.

My way of making feature requests is use build2, find difficulties wort to be fixed, find good solution - make feature request. I understand that from strategic point of view my feature requests are stupid. I just think I need to make build2 more easy to use because each difficulties will force people go out (exactly as absence of required feature).

Also I thing we do not need to make build2 perfect from first time. If something can be fixed by hack - just  worry about this hack can be removed later (hack need to extend your architecture not crash it). Look for cmake - they make easy to use tool (compare with autotools) and now there is "modern cmake" which is not perfect too but better then old cmake way (old is legacy but supported). IMHO, it is impossible to make perfect tool in one time. It's evolution. So You can make hack then look for better solution (which can be burned from solving other task) or You can think about solution (without providing fix) and people will go out. IMHO, the only working way is to make something then make  it better. In this case add hack, then when better solution become available implement it and provide legacy hack support for some time. In both my feature requests (.gitignore and this one) You will never even need to fix it because it will not interfere something.

**Comment by boris-kolpackov** [comment]
A quick progress report: the `0.10.0` release allows more granular source file extension specification, for example:

```
$ bdep new -t lib -l c++,hxx=h,cxx=cpp  libhello # Use .h and .cpp.
```

Combined with the `--options-file` mechanism, one can do something like this:

```
$ cat <<EOF >bdep-new.options
--lang c++,hxx=h,cxx=cpp
EOF

$ bdep new --options-file bdep-new.options -t lib -l c++ libhello
```

Of course, other aspects of `bdep-new` can be customized this way as well, for example:

```
$ cat <<EOF >bdep-new.options
--type exe,unit-tests,alt-naming
--type lib,no-version,alt-naming
--lang c++,hxx=h,cxx=cpp
EOF
```

Finally, the latest staged version (so it will be available in the `0.11.0` release) adds the `--post-hook` option that allows one to run arbitrary customization commands after the project is created. For example:

```
$ bdep new --post-hook "echo .idea/ >>.gitignore" hello
```

Another piece that we will try to finish before `0.11.0` is support for configuration files (so that one doesn't have to keep specifying `--options-file ...`). The idea is to reuse the same options file mechanism by searching for some well known file names in some well known locations. While the details still need to be worked out, I like @Klaim 's idea of searching in outer directories (potentially starting with the project's directory itself so one can have project-specific `bdep-new` configuration).

**Comment by boris-kolpackov** [comment]
Adding a comment from the `build2` slack channel with some updates:

While we thought of templates and might still support them in the future, I believe there is already (especially if you use the staged version or the 0.12.0 release once it's out) enough functionality to do pretty much that and more. Specifically, in the 0.12.0 release (and already on stage) `bdep-new` can create a project in a directory that already contains files provided they don't clash with what `bdep-new` tries to create. The idea here is to support the "create project on GitHub/Lab/etc clone it and then run `bdep-new`" kind of workflow (and these services also provide templating systems). In this case, `bdep-new` will even use information from `LICENSE` and `README.md` when generating `manifest`.  The second bit is the pre/post-creation hooks (`--pre-hook`/`--post-hook`) which allow you to run arbitrary commands before/after `bdep-new` creates its files. Finally, add to that support for default options files (aka config files; also already available on stage) and you can arrange everything in a way that running just `bdep new` creates a project exactly as you want it.

**Comment by boris-kolpackov** [comment]
In the `0.12.0` release we've added support for default options files (aka tool config files). For details see the `DEFAULT OPTIONS FILES` section in `b(1)`, `bdep-<command>(1)` and `bpkg-<command>(1)`.

Also, in this release `bdep-new` supports creation from existing files as well as the `--pre-hook` option. While this doesn't completely eliminate the need for the interactive mode, it probably reduces it somewhat. So I am going to close this but we will keep an eye on whether interactive mode is requested in the future.
